File: dotnet/Services/CurrencyConfig.cs
using System.Collections.Generic;
using System.Globalization;

namespace GlobalPayments.Services
{
    /// <summary>
    /// Currency Configuration
    /// Manages currency metadata and formatting rules
    /// </summary>
    public class CurrencyConfig
    {
        public class Currency
        {
            public string Code { get; set; }
            public string Name { get; set; }
            public string Symbol { get; set; }
            public string SymbolPosition { get; set; }
            public int Decimals { get; set; }
            public string DecimalSeparator { get; set; }
            public string ThousandsSeparator { get; set; }
            public string Country { get; set; }
            public string Flag { get; set; }

            public Currency(string code, string name, string symbol, string symbolPosition,
                          int decimals, string decimalSeparator, string thousandsSeparator,
                          string country, string flag)
            {
                Code = code;
                Name = name;
                Symbol = symbol;
                SymbolPosition = symbolPosition;
                Decimals = decimals;
                DecimalSeparator = decimalSeparator;
                ThousandsSeparator = thousandsSeparator;
                Country = country;
                Flag = flag;
            }
        }

        private static readonly Dictionary<string, Currency> Currencies = new()
        {
            { "USD", new Currency("USD", "US Dollar", "$", "before", 2, ".", ",", "US", "🇺🇸") },
            { "EUR", new Currency("EUR", "Euro", "€", "after", 2, ",", ".", "GB", "🇪🇺") }, // GP API uses GB for EUR
            { "GBP", new Currency("GBP", "British Pound", "£", "before", 2, ".", ",", "GB", "🇬🇧") },
            { "CAD", new Currency("CAD", "Canadian Dollar", "C$", "before", 2, ".", ",", "CA", "🇨🇦") },
            { "AUD", new Currency("AUD", "Australian Dollar", "A$", "before", 2, ".", ",", "AU", "🇦🇺") },
            { "JPY", new Currency("JPY", "Japanese Yen", "¥", "before", 0, "", ",", "JP", "🇯🇵") }
        };

        /// <summary>
        /// Get all supported currencies
        /// </summary>
        public static Dictionary<string, Currency> GetAllCurrencies()
        {
            return new Dictionary<string, Currency>(Currencies);
        }

        /// <summary>
        /// Get currency configuration by code
        /// </summary>
        public static Currency GetCurrency(string code)
        {
            return Currencies.GetValueOrDefault(code.ToUpper());
        }

        /// <summary>
        /// Check if currency is supported
        /// </summary>
        public static bool IsSupported(string code)
        {
            return Currencies.ContainsKey(code.ToUpper());
        }

        /// <summary>
        /// Get country code for currency (for GP API configuration)
        /// </summary>
        public static string GetCountryCode(string currencyCode)
        {
            var currency = GetCurrency(currencyCode);
            return currency?.Country ?? "US";
        }

        /// <summary>
        /// Format amount according to currency rules
        /// </summary>
        public static string FormatAmount(decimal amount, string currencyCode)
        {
            var currency = GetCurrency(currencyCode);
            if (currency == null)
            {
                return amount.ToString("0.00");
            }

            // Format with proper decimal places
            string formatted = amount.ToString($"F{currency.Decimals}");

            // Replace decimal separator
            if (!string.IsNullOrEmpty(currency.DecimalSeparator) && currency.DecimalSeparator != ".")
            {
                formatted = formatted.Replace(".", currency.DecimalSeparator);
            }

            // Add thousands separator
            var parts = formatted.Split(new[] { '.', ',' });
            if (parts.Length > 0)
            {
                string integerPart = parts[0];
                var result = new System.Text.StringBuilder();
                int count = 0;

                for (int i = integerPart.Length - 1; i >= 0; i--)
                {
                    if (count > 0 && count % 3 == 0)
                    {
                        result.Insert(0, currency.ThousandsSeparator);
                    }
                    result.Insert(0, integerPart[i]);
                    count++;
                }

                if (currency.Decimals > 0 && parts.Length > 1)
                {
                    result.Append(currency.DecimalSeparator).Append(parts[1]);
                }

                formatted = result.ToString();
            }

            // Add currency symbol
            if (currency.SymbolPosition == "before")
            {
                return currency.Symbol + formatted;
            }
            else
            {
                return formatted + " " + currency.Symbol;
            }
        }

        /// <summary>
        /// Get supported currency codes
        /// </summary>
        public static IEnumerable<string> GetSupportedCodes()
        {
            return Currencies.Keys;
        }

        /// <summary>
        /// Validate and normalize currency code
        /// </summary>
        public static string ValidateCurrency(string code)
        {
            if (string.IsNullOrWhiteSpace(code))
            {
                return "USD";
            }

            string upperCode = code.ToUpper().Trim();
            return IsSupported(upperCode) ? upperCode : "USD";
        }
    }
}

File: nodejs/services/CurrencyConfig.js
/**
 * Currency Configuration
 * Manages currency metadata and formatting rules
 */

class CurrencyConfig {
  static currencies = {
    USD: {
      code: 'USD',
      name: 'US Dollar',
      symbol: '$',
      symbolPosition: 'before',
      decimals: 2,
      decimalSeparator: '.',
      thousandsSeparator: ',',
      country: 'US',
      flag: '🇺🇸'
    },
    EUR: {
      code: 'EUR',
      name: 'Euro',
      symbol: '€',
      symbolPosition: 'after',
      decimals: 2,
      decimalSeparator: ',',
      thousandsSeparator: '.',
      country: 'GB', // GP API uses GB for EUR processing
      flag: '🇪🇺'
    },
    GBP: {
      code: 'GBP',
      name: 'British Pound',
      symbol: '£',
      symbolPosition: 'before',
      decimals: 2,
      decimalSeparator: '.',
      thousandsSeparator: ',',
      country: 'GB',
      flag: '🇬🇧'
    },
    CAD: {
      code: 'CAD',
      name: 'Canadian Dollar',
      symbol: 'C$',
      symbolPosition: 'before',
      decimals: 2,
      decimalSeparator: '.',
      thousandsSeparator: ',',
      country: 'CA',
      flag: '🇨🇦'
    },
    AUD: {
      code: 'AUD',
      name: 'Australian Dollar',
      symbol: 'A$',
      symbolPosition: 'before',
      decimals: 2,
      decimalSeparator: '.',
      thousandsSeparator: ',',
      country: 'AU',
      flag: '🇦🇺'
    },
    JPY: {
      code: 'JPY',
      name: 'Japanese Yen',
      symbol: '¥',
      symbolPosition: 'before',
      decimals: 0,
      decimalSeparator: '',
      thousandsSeparator: ',',
      country: 'JP',
      flag: '🇯🇵'
    }
  };

  /**
   * Get all supported currencies
   * @returns {Object} All currency configurations
   */
  static getAllCurrencies() {
    return this.currencies;
  }

  /**
   * Get currency configuration by code
   * @param {string} code - Currency code (USD, EUR, etc.)
   * @returns {Object|null} Currency configuration or null
   */
  static getCurrency(code) {
    const upperCode = code.toUpperCase();
    return this.currencies[upperCode] || null;
  }

  /**
   * Check if currency is supported
   * @param {string} code - Currency code
   * @returns {boolean} True if supported
   */
  static isSupported(code) {
    return !!this.currencies[code.toUpperCase()];
  }

  /**
   * Get country code for currency (for GP API configuration)
   * @param {string} currencyCode - Currency code
   * @returns {string} Country code
   */
  static getCountryCode(currencyCode) {
    const currency = this.getCurrency(currencyCode);
    return currency ? currency.country : 'US';
  }

  /**
   * Format amount according to currency rules
   * @param {number} amount - Amount to format
   * @param {string} currencyCode - Currency code
   * @returns {string} Formatted amount with currency symbol
   */
  static formatAmount(amount, currencyCode) {
    const currency = this.getCurrency(currencyCode);
    if (!currency) {
      return amount.toFixed(2);
    }

    // Format with proper decimal places and separators
    const parts = amount.toFixed(currency.decimals).split('.');
    const integerPart = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, currency.thousandsSeparator);
    const decimalPart = parts[1];

    let formatted = integerPart;
    if (currency.decimals > 0 && decimalPart) {
      formatted += currency.decimalSeparator + decimalPart;
    }

    // Add currency symbol
    if (currency.symbolPosition === 'before') {
      return currency.symbol + formatted;
    } else {
      return formatted + ' ' + currency.symbol;
    }
  }

  /**
   * Get supported currency codes
   * @returns {Array<string>} Array of currency codes
   */
  static getSupportedCodes() {
    return Object.keys(this.currencies);
  }

  /**
   * Validate and normalize currency code
   * @param {string|null} code - Currency code
   * @returns {string} Valid currency code (defaults to USD)
   */
  static validateCurrency(code) {
    if (!code) {
      return 'USD';
    }

    const upperCode = code.toUpperCase().trim();
    return this.isSupported(upperCode) ? upperCode : 'USD';
  }
}

export default CurrencyConfig;
